File: python/Day21/day21.py
import re, hashlib, itertools, numpy as np
from collections import *


def array_to_pattern_string(nparray):
    return '/'.join([''.join(row) for row in nparray])


def get_variations(ip):
    ip = [[c for c in row] for row in ip.split('/')]
    array = np.asarray(ip)
    variations = set()
    for rot in range(4):
        variations.update([array_to_pattern_string(np.rot90(array, rot)),
                           array_to_pattern_string(np.fliplr(np.rot90(array, rot))),
                           array_to_pattern_string(np.flipud(np.rot90(array, rot)))])

    return variations


grid = [['.', '#', '.'], ['.', '.', '#'], ['#', '#', '#']]
inpt = open('day21.in').read().strip()
data = inpt.splitlines()

artist_book = {2: {}, 3: {}}
for line in data:
    input_pattern, output_pattern = line.split(' => ')
    output_pattern = tuple(output_pattern.split('/'))
    for variation in get_variations(input_pattern):
        artist_book[len(output_pattern) - 1][variation] = output_pattern

for _ in range(18):
    if len(grid) % 2 == 0:
        group_size = 2
    else:
        group_size = 3

    groups = len(grid) // group_size
    new_grid = [[] for _ in range(groups * (group_size + 1))]
    current_book = artist_book[group_size]

    for r in range(groups):
        for c in range(groups):
            section = [row[c * group_size:(c + 1) * group_size] for row in grid[r * group_size: (r + 1) * group_size]]
            output = current_book['/'.join([''.join(row) for row in section])]
            for nr in range(group_size + 1):
                new_grid[nr + (group_size + 1) * r].extend(output[nr])
    grid = new_grid

print(sum([row.count('#') for row in grid]))

File: kotlin/src/main/AoC2017/day21.kt
package main.AoC2017

import java.io.File


fun main(args: Array<String>){
    
    val input = File("inputs/2017/day21.in").readLines()
    val artistBook = mutableMapOf<String, List<List<Int>>>()
    var grid = patternStringToMatrix(".#./..#/###")
    input.map { line -> 
            val (inputPattern, outputPattern) = line.split(" => ")
            val outputMatrix = patternStringToMatrix(outputPattern)
            val variations = getAllMatrixVariations(patternStringToMatrix(inputPattern))
            for(variation in variations){
                artistBook[matrixToPatternString(variation)] = outputMatrix
            }
    }

    repeat(18){
        val groupSize = if(grid.size % 2 == 0) 2 else 3
        val groups = grid.size / groupSize
        val newGrid = List(groups * (groupSize + 1)){_ -> mutableListOf<Int>()}
        for(r in 0 until groups){
            for(c in 0 until groups){
                val currentGroup = grid
                        .subList(r * groupSize,(r + 1) * groupSize)
                        .map { it.subList(c * groupSize, (c + 1) * groupSize) }
                val output = artistBook.getValue(matrixToPatternString(currentGroup))
                for(i in output.indices){
                    newGrid[i + (groupSize + 1) * r].addAll(output[i])
                }
            }
        }
        grid = newGrid.map { it.toList() }

    }
    println(grid.sumBy { it.sum() })
 
}



fun rot90(matrix: List<List<Int>>): List<List<Int>>{
    return transpose(matrix.asReversed())
}

fun transpose(matrix: List<List<Int>>): List<List<Int>>{
    val size = matrix.size
    val newList = MutableList(size){ _ -> MutableList<Int>(size){ _ -> 0} }
    for(i in 0 until size){
        for(j in 0 until size){
            newList[j][i] = matrix[i][j]
        }
    }
    return newList.map { it.toList() }.toList()
}

fun flipUD(matrix: List<List<Int>>): List<List<Int>>{
    val newList = matrix.map { it.toMutableList() }
    for(r in 0 until matrix.size){
        for(c in 0 until matrix.size){
            newList[matrix.size - 1 - r][c] = matrix[r][c]
        }
    }
    return newList.map { it.toList() }
}

fun flipLR(matrix: List<List<Int>>): List<List<Int>>{
    val newList = matrix.map { it.toMutableList() }
    for(r in 0 until matrix.size){
        for(c in 0 until matrix.size){
            newList[r][matrix.size - 1 - c] = matrix[r][c]
        }
    }
    return newList.map { it.toList() }
}

fun getAllMatrixVariations(matrix: List<List<Int>>): Set<List<List<Int>>>{
    val set = mutableSetOf(matrix)
    var current = matrix
    for(i in 0 until 4){
        set.add(current)
        set.add(flipLR(current))
        set.add(flipUD(current))
        current = rot90(current)
    }
    return set
}

fun matrixToPatternString(matrix: List<List<Int>>): String{
    return matrix.joinToString("/") { it.map { i -> if(i == 1) '#' else '.' }.joinToString("") }
}

fun patternStringToMatrix(pattern: String): List<List<Int>>{
    return pattern.split("/").map { it.map { if (it == '#') 1 else 0} }
}